src/Control/Carrier/Cull/Church.hs

```diff
@@ -7,28 +7,30 @@ module Control.Carrier.Cull.Church
 , runCullA
 , runCullM
 , CullC(..)
-  -- * Cull effect
-, module Control.Effect.Cull
-  -- * NonDet effects
-, module Control.Effect.NonDet
+  -- * Cull effects
+, module X
```

Comment:
Hmm, it looks like using the `X` trick decreases the quality of the Haddocks. Before, the `Cull effect` header just linked to that module:

<img width="845" alt="Screen Shot 2019-10-19 at 2 16 23 PM" src="https://user-images.githubusercontent.com/2498/67149615-45f39000-f27b-11e9-88a5-a9c2eac5a43a.png">

Afterwards, we embed _everything_ about the aforementioned modules, up to and including a definition of `Alternative`:

<img width="647" alt="Screen Shot 2019-10-19 at 2 19 03 PM" src="https://user-images.githubusercontent.com/2498/67149628-79361f00-f27b-11e9-8e0b-0c62f6960978.png">

I think that’s a sufficient step back in readability and brevity (every such page becomes very much longer) that we should avoid the `X` trick, or figure out some Haddock pragma to prevent it. I really like the effect/carrier distinction, and that distinction becomes very hard to see externally if we Haddock-inline all the definitions exported from the effect—there’s nothing that indicates to the user the borders of this distinction that we’ve tried so hard to make. What do you think?

Reply:
Yeah, I noticed this too. There isn’t a haddock pragma to prevent it that I could find.

I honestly think it’s an improvement that the haddocks list everything you get from the import, but I get what you mean, and I wouldn’t mind at all if it were better separated, or hidden by default, or something—but I don’t know how to do that.

To be precise, the direct cause is that we’re hiding some of the re-exported module(s). We could avoid this by defining the effect datatypes in _another_ module, and having the `Control.Effect.Blah` module import that and re-export only the type. (IMO that would be an improvement since in the current setup, importing `Control.Effect.Reader` to use `ask` also brings in e.g. `Local`—the exact same problem as this PR was designed to resolve for the carrier modules.

I’d be … ok, I guess … with adding `Control.Effect.Blah.Type` modules, or something like that. We can even use `{-# OPTIONS_HADDOCK home #-}` to make links resolve to `Control.Effect.Blah` instead of the internal one. We’d still have to add `Control.Effect.Blah.Type` to the exported modules list, tho, or else no one else could define carriers; and that’s going to make the table of contents for the package _really_ lengthy.

Reply:
Follow-up: we decided to just re-export all the constructors and hide them on the user side if necessary.